src/Lecture2.hs

```diff
@@ -74,7 +78,12 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a,[a])
+removeAt n l
+  | null l = (Nothing,[])
+  | n < 0 = (Nothing, l)
+  | null (drop n l) = (Nothing, l) 
+  | otherwise = (Just (l !! n), take n l ++ drop (n+1) l)
```

Comment:
That looks correct and neat 👍🏻 

Potential performance optimization is to avoid traversing the list prefix multiple times (3 times to be more precise: using `take`, `drop` and `!!`) and extract the needed element in one go.

For this, you can use a standard function [splitAt](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:splitAt) instead of the last two guards.

Reply:
I have a solution with splitAt now. One win is that the 3rd and 4th guard is now sharing part of it. Still, splitAt is in fact a take and drop as well, right? Does it win a lot of performance this way? Or is my new solution not what you had in mind?

Reply:
`splitAt` is semantically equivalent to `take` + `drop` but it's implemented in a more performant way using only a single traversal. You can check the source code of `splitAt`:

```haskell
splitAt n ls
  | n <= 0 = ([], ls)
  | otherwise          = splitAt' n ls
    where
        splitAt' :: Int -> [a] -> ([a], [a])
        splitAt' _  []     = ([], [])
        splitAt' 1  (x:xs) = ([x], xs)
        splitAt' m  (x:xs) = (x:xs', xs'')
          where
            (xs', xs'') = splitAt' (m - 1) xs
```

So, it's still better than doing both `take` and `drop`. You can implement a manual recursion in `removeAt` in a similar way to `splitAt` to avoid using any standard functions at all. But standard functions are already optimized for performance and laziness. And it's hard to tell which solution is "better".